text/0812-tracked-built-ins.md

```diff
@@ -0,0 +1,287 @@
+mistakenly believe this was an error, and remove this statement, causing
+bugs. It's not a very google-able pattern either, which makes it more
+difficult to learn and to teach. Finally, it's fairly easy this way for the
+state to get out of sync with the rest of the system if a developer forgets
+to add the extra re-set after mutating the value.
+
+Over time, it's become clear that this pattern is in actuality an antipattern
+for these reasons, and something we should begin to discourage in general.
+
+[RFC #669 "Tracked Storage Primitives"][rfc-0669] introduced low level primitives
+that allow to implement tracked versions of JavaScript's built-in classes
+in an addon like [tracked-built-ins](https://github.com/tracked-tools/tracked-built-ins):
+
+```ts
+import {
+  TrackedObject,
+  TrackedArray,
+  TrackedMap,
+  TrackedSet,
+  TrackedWeakMap,
+  TrackedWeakSet,
+} from 'tracked-built-ins';
+
+class Foo {
+  @tracked value = 123;
+
+  obj = new TrackedObject<{[string: number]}>({ foo: 123 });
+  arr = new TrackedArray<string, number>(['foo', 123]);
+  map = new TrackedMap<string, number>(['foo', 123]);
+  set = new TrackedSet<string, number>(['foo', 123]);
+  weakMap = new TrackedWeakMap<{[object: number]}>([[{}, 456]]);
+  weakSet = new TrackedWeakSet<string, number>(['foo', 123]);
+}
+```
+
+or using `tracked()` decorator:
+
+```ts
+import { tracked } from 'tracked-built-ins';
+
+class Foo {
+  @tracked value = 123;
+
+  obj = tracked<{[string: number]}>({ foo: 123 });
+  arr = tracked<string, number>(['foo', 123]);
+  map = tracked<string, number>(new Map(['foo', 123]));
+  set = tracked<string, number>(new Set(['foo', 123]));
+  weakMap = tracked<{[string: number]}>(new WeakMap([[{}, 456]]));
+  weakSet = tracked<string, number>(new WeakSet(['foo', 123]));
+}
+```
+
+## Detailed design
+
+The necessary changes to `ember-cli` are relatively small since we only need
+to add the dependency to the `app` blueprint, and the `addon` blueprint will
+inherit it automatically.
+
+This has the advantage (over including it as an implicit dependency), that
+apps and addons that don't want to use it for some reason can opt out by
+removing the dependency from their `package.json` file.
+
+## How we teach this
+
+The following guide should be added to the Autotracking In-Depth guide in the official guides, and the
+sections on [POJOs][guides-pojos] and [Array][guides-arrays] should be removed.
+
+### Tracked data structures
+
+Generally, you should try to create classes with their tracked properties enumerated
+and decorated with `@tracked`, instead of relying on dynamically created objects.
+In some cases however, there could be a prohibitive number of possible properties,
+or there could be no way to know them in advance.
+In this case, you should use tracked versions of JavaScript's built-in
+Plain Old JavaScript Object (POJO), array or keyed collection (Maps, Sets, WeakMaps, WeakSets)
```

Comment:
should we explain why certain structures are or aren't supported?
Like, I have some projects that make heavy use of [Uint32Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint32Array), I'm sure other folks have other stuff they'd like tracked versions of (or a way to optimally track them)